### PR TITLE
HPCC-29768 Protect updates with a dali lock when fetching a git repo

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
@jakesmith for initial review.  See the jira for details of the change.  I will add some more changes in a different PR to i) pre-cache the repo and ii) delete old lock files.

File: helm/hpcc/values.schema.json [hunk 1]
```diff
@@ -1421,6 +1421,10 @@
           "type": "string",
           "description": "The default repo version used if not supplied for the defaultRepo"
         },
+        "guardGitUpdates": {
+          "type": "boolean",
+          "description": "If enabled all updates of the git repositories is protected by holding a lock in dali"
```
Comment (jakesmith) [inline helm/hpcc/values.schema.json:1426]: trivial: .. 'are' protected?

File: helm/hpcc/templates/eclccserver.yaml [hunk 2]
```diff
@@ -27,6 +27,7 @@ Pass in dict with root and me
 */}}
 {{- define "hpcc.eclccServerConfigMap" -}}
 {{- $compileJobName := printf "compile-job-_HPCC_JOBNAME_" }}
+{{- $gitPlane := .me.gitPlane | default (include "hpcc.getDefaultGitPlane" .root) -}}
```
Comment (jakesmith) [inline helm/hpcc/templates/eclccserver.yaml:30]: trivial: I notice the pre-existing line above didn't/doesn't have a whitespace chomping char , but this new line ahead of the content does.. Don't think it's problematic, perhaps make both lines same.

File: ecl/eclcc/eclcc.hpp [hunk 3]
```diff
@@ -96,6 +96,7 @@ const char * const helpText[] = {
     "?!  --fastsyntax  Delay expanding functions when parsing.  May speed up processing for some queries",
     "?   --fetchrepos  Automatically download missing repositories associated with dependencies",
     "!   --gituser=x   Which user should be used for accessing git repositories (for servers)",
+    "!   --gitlock=key The dali key (e.g. plane name) that should be used to project updates to git repositories",
```
Comment (jakesmith) [inline ecl/eclcc/eclcc.hpp:99]: typo: project -> protect.

File: ecl/eclccserver/eclccserver.cpp [hunk 4]
```diff
@@ -235,6 +235,45 @@ static bool getHomeFolder(StringBuffer & homepath)
...
+                        if (isalnum(*cur))
+                            gitLockKey.append(*cur);
+                        else
+                            gitLockKey.append("_");
+                    }
+                }
+            }
+
+            if (!gitLockKey.isEmpty())
+                guardGitUpdates = true;
```
Comment (jakesmith) [inline ecl/eclccserver/eclccserver.cpp:272]: trivial: any point to non-empty gitLockKey and a boolean? i.e. could just test if (!gitLockKey.isEmpty()) below:

**Review by jakesmith — COMMENTED** [review]
@ghalliday - looks good afaics, a few trivial comments.

File: ecl/eclccserver/eclccserver.cpp (hunk 4 above)
Comment (ghalliday) [reply ecl/eclccserver/eclccserver.cpp:272]: Could do, but it is arguably slightly easier to read the code below.  I'm not sure it is worth changing.

File: helm/hpcc/templates/eclccserver.yaml (hunk 2 above)
Comment (ghalliday) [reply helm/hpcc/templates/eclccserver.yaml:30]: I will change to match

**Review by jakesmith — APPROVED** [review]
@ghalliday - looks good, ready to squash.

**Pull request merged**